vortex-array/src/validity.rs

```diff
@@ -376,6 +376,37 @@ impl Validity {
     pub fn is_array(&self) -> bool {
         matches!(self, Validity::Array(_))
     }
+
+    /// Given monotonically increasing `row_indices` in [0, n_rows], returns the
+    /// count of valid elements up to each row_index.
+    ///
+    /// This is O(n_rows).
+    pub fn value_indices_for_rows(&self, row_indices: &[usize]) -> VortexResult<Vec<usize>> {
+        let value_indices = match self {
+            Validity::NonNullable | Validity::AllValid => row_indices.to_vec(),
+            Validity::AllInvalid => vec![0; row_indices.len()],
+            Validity::Array(is_valid) => {
+                let bool_array = is_valid.to_bool()?;
+                let mut bool_iter = bool_array.boolean_buffer().iter();
+                let mut value_indices = Vec::with_capacity(row_indices.len());
+                let mut value_idx = 0;
+                let mut row_idx = 0;
+                for &next_row_idx in row_indices {
+                    while row_idx < next_row_idx {
+                        row_idx += 1;
+                        value_idx += bool_iter
+                            .next()
+                            .ok_or_else(|| vortex_err!("Row indices exceed array length"))?
+                            as usize;
+                    }
+                    value_indices.push(value_idx);
+                }
```

Comment:
```suggestion
                let mut bool_iter = bool_array.boolean_buffer().set_indices().peekable();
                let mut value_indices = Vec::with_capacity(row_indices.len());
                let mut value_idx = 0;
                for &next_row_idx in row_indices {
                    if let Some(&next_set_idx) = bool_iter.peek() {
                        if next_set_idx < next_row_idx {
                            value_idx += bool_iter
                                .by_ref()
                                .peeking_take_while(|&set_idx| set_idx < next_row_idx)
                                .count();
                        }
                    }
                    value_indices.push(value_idx);
                }
```
I think something like this would work better where we only iterate set indices. You very rarely want to iterate every bit.

Reply:
I actually think the original will be more performant:
1. set_indices is linearly iterating over every bit to begin with
2. There are a lot fewer Options and complex iterator types in the original. Some of these will probably get zero-costed away, but I'd guess not all.